README.md
# ReCharge API Client

[![Build Status](https://travis-ci.org/ScreenStaring/recharge-api.svg?branch=master)](https://travis-ci.org/ScreenStaring/recharge-api)

Ruby client for [ReCharge Payments'](https://rechargepayments.com/developers)
recurring payments API for Shopify.

## Installation

Ruby gems:

    gem install recharge-api

Bundler:

    gem "recharge-api", :require => "recharge"

## Usage

An API key is required. The key can be set via `ReCharge.api_key` or via the `RECHARGE_API_KEY`
environment variable.

```rb
require "recharge"

ReCharge.api_key = "YOUR_KEY"  # Can also use Recharge
data = {
  :address_id => 123234321,
  :customer_id => 565728,
  # ... more stuff
  :next_charge_scheduled_at => Time.new,
  :properties => {
    :name => "size",
    :value => "medium"
  }
}

subscription = ReCharge::Subscription.create(data)
subscription.address_id = 454343
subscription.save

# Or
ReCharge::Subscription.update(id, data)

subscription = ReCharge::Subscription.new(data)
subscription.save

order1 = ReCharge::Order.get(123123)
order1.line_items.each do |li|
  p li.title
  p li.quantity
end

order2 = ReCharge::Order.get(453321)
p "Different" if order1 != order2

JSON.dump(order2.to_h)

customers = ReCharge::Customer.list(:page => 10, :limit => 50)
customers.each do |customer|
  addresses = ReCharge::Customer.addresses(customer.id)
  # ...
end
```

For complete documentation refer to the API docs: http://rdoc.info/gems/recharge-api

## Rake Tasks for Webhook Management

Add the following to your `Rakefile`:

```rb
require "recharge/tasks"
```

This will add the following tasks:

  * `recharge:hook:create` - create webhook `HOOK` to be sent to `CALLBACK`
  * `recharge:hooks:delete` - delete the webhook(s) given by `ID`
  * `recharge:hooks:delete_all` - delete all webhooks
  * `recharge:hooks:list` - list webhooks

All tasks require `RECHARGE_API_KEY` be set.

For example, to create a hook run the following:

```
rake recharge:hooks:create RECHARGE_API_KEY=YOURKEY HOOK=subscription/created CALLBACK=https://example.com/callback
```

## License

Released under the MIT License: www.opensource.org/licenses/MIT

---

Made by [ScreenStaring](http://screenstaring.com)


Changes
2020-11-27  v0.0.3
--------------------
* Fix HTTP requests so that they're thread-safe
* Fix Discount#value; it's a Float not an Integer

2020-05-05  v0.0.2
--------------------
* Fix for HTTP 204 responses
* Add support for Metafield
* Add limited support for Product and Discount
* Add additional properties to Subscription and Charge
* Add support for setting the API key via the RECHARGE_API_KEY env var


spec/order_spec.rb
require "spec_helper"

RSpec.describe Recharge::Order do
  subject { described_class }

  it { is_expected.to be_a(Recharge::HTTPRequest::Get) }
  it { is_expected.to be_a(Recharge::HTTPRequest::Count) }
  it { is_expected.to be_a(Recharge::HTTPRequest::List) }
  it { is_expected.to define_const("COLLECTION").set_to("orders") }
  it { is_expected.to define_const("SINGLE").set_to("order") }
  it { is_expected.to define_const("PATH").set_to("/orders") }

  describe ".new" do
    it "instantiates an order object with the given attributes" do
      data = {
        "id" => 1,
        "customer_id" => 2,
        "address_id" => 3,
        "charge_id" => 4,
        "transaction_id" => "TID",
        "shopify_order_id" => "5",
        "shopify_order_number" => 6,
        "created_at" => "2017-01-01",
        "updated_at" => "2017-01-02",
        "scheduled_at" => "2017-01-02",
        "processed_at" => "2017-01-03",
        "status" => "SUCCESS",
        "charge_status" => "SUCCESS",
        "type" => "CHECKOUT",
        "first_name" => "sshaw",
        "last_name" => "X",
        "email" => "s@example.com",
        "payment_processor" => "4 much",
        "address_is_active" => 1,
        "is_prepaid" => false,
        "line_items" => [],
        "total_price" => 100.0,
        "shipping_address" => [],
        "billing_address" => []
      }

      order = described_class.new(data)
      expect(order.to_h).to eq data
    end
  end

  describe ".get" do
    it_behaves_like "a method that requires an id"

    it "makes a GET request to the orders endpoint for the given id" do
      id = 123
      order = described_class.new(:id => id)
      expect(described_class).to receive(:GET)
      	                           .with("/orders/#{id}")
                                   .and_return("order" => { "id" => id })

      expect(described_class.get(id)).to eq order
    end
  end

  describe ".update_shopify_variant" do
    it "makes a POST request to update_shopify_variant" do
      old_variant_id = 999
      new_variant_id = 123
      order = described_class.new(:id => 123)

      expect(described_class).to receive(:POST)
      	                           .with("/orders/#{order.id}/update_shopify_variant/#{old_variant_id}", :new_shopify_variant_id => 123)
                                   .and_return("order" => { "id" => new_variant_id })


      expect(described_class.update_shopify_variant(order.id, old_variant_id, new_variant_id)).to eq described_class.new(:id => 123)
    end
  end

  describe ".change_date" do
    it "makes a POST request to change_date with the given order id" do
      order = described_class.new(:id => 1)
      time = Time.new

      expect(described_class).to receive(:POST)
      	                           .with("/orders/#{order.id}/change_date", :shipping_date => format_time(time))
                                   .and_return("order" => { "id" => 1 })


      expect(described_class.change_date(order.id, time)).to eq order
    end
  end

  describe ".count" do
    before do
      @path = "/orders/count"
      @retval = { "count" => 1 }
    end

    it "makes a GET request to count and returns the count" do
      expect(described_class).to receive(:GET)
      	                           .with(@path, nil)
                                   .and_return(@retval)

      expect(described_class.count).to eq 1
    end

    %i[created_at_max created_at_min date_min date_max].each do |param|
      it "filters the count on #{param}" do
        time = Time.now

        expect(described_class).to receive(:GET)
      	                             .with(@path, param => format_time(time))
                                     .and_return(@retval).twice

        expect(described_class.count(param => time)).to eq 1
        expect(described_class.count(param => format_time(time))).to eq 1
      end
    end
  end
end


spec/charge_spec.rb
require "spec_helper"

RSpec.describe Recharge::Charge do
  subject { described_class }

  it { is_expected.to be_a(Recharge::HTTPRequest::Get) }
  it { is_expected.to be_a(Recharge::HTTPRequest::Count) }
  it { is_expected.to be_a(Recharge::HTTPRequest::List) }

  it { is_expected.to define_const("PATH").set_to("/charges") }
  it { is_expected.to define_const("SINGLE").set_to("charge") }
  it { is_expected.to define_const("COLLECTION").set_to("charges") }

  describe ".skip" do
    it "makes a POST request to the skip endpoint with the given id and subscription id" do
      sub_id = 999
      charge = described_class.new(:id => 123)

      expect(described_class).to receive(:POST)
      	                           .with("/charges/#{charge.id}/skip", :subscription_id => sub_id)
                                   .and_return("charge" => { "id" => charge.id })

      expect(described_class.skip(charge.id, sub_id)).to eq charge
    end
  end

  describe ".change_next_charge_date" do
    it "makes a POST request to the change_next_charge_date endpoint with the given id and date" do
      date = "2017-01-01"
      charge = described_class.new(:id => 123)

      expect(described_class).to receive(:POST)
      	                           .with("/charges/#{charge.id}/change_next_charge_date", :next_charge_date => date)
                                   .and_return("charge" => { "id" => charge.id })

      expect(described_class.change_next_charge_date(charge.id, date)).to eq charge
    end

    it "converts date/time instances" do
      TIME_INSTANCES.each do |time|
        expect(described_class).to receive(:POST)
      	                             .with("/charges/1/change_next_charge_date", :next_charge_date => format_time(time))
                                     .and_return("charge" => { "id" => 1 })

        described_class.change_next_charge_date(1, time)
      end
    end
  end

  describe ".list" do
    %i[date_min date_max].each do |param|
      it "filters on #{param}" do
        # set has_uncommited_changes as charge.to_h => { "has_uncommited_changes" => nil }
        # and Charge.new("has_uncommited_changes" => nil).has_uncommited_changes == false
        charge = described_class.new(:id => 1, :has_uncommited_changes => false)

        TIME_INSTANCES.each do |time|
          expect(described_class).to receive(:GET)
                                       .with("/charges", param => format_time(time))
                                       .and_return("charges" => [charge.to_h])

          expect(described_class.list(param => time)).to eq [charge]

          expect(described_class).to receive(:GET)
                                       .with("/charges", param => format_time(time))
                                       .and_return("charges" => [charge.to_h])

          expect(described_class.list(param => format_time(time))).to eq [charge]

        end
      end
    end
  end

  describe ".new" do
    it "instantiates a charge object with the given attributes" do
      data = {
        "id" => 1,
        "address_id" => 2,
        "analytics_data" => { "foo" => 999 },
        "billing_address" => {
          "city" => "NYC",
          "address1" => "555 5th ave",
          "address2" => "10th floor",
          "company"  => "foo",
          "country"  => "USA",
          "first_name" => "s",
          "last_name" => "shaw",
          "phone" => "555-555-1212",
          "province" => "BC",
          "zip" => "10000"
        },
        "shipping_address" => {
          "city" => "LA",
          "address1" => "123 Sepulveda Blvd.",
          "address2" => "10th floor",
          "company"  => "bar",
          "country"  => "USA",
          "first_name" => "s",
          "last_name" => "shaw",
          "phone" => "555-555-9999",
          "province" => "BC",
          "zip" => "90210"
        },
        "client_details" => {
          "browser_ip" => "127.0.0.1",
          "user_agent" => "Konquer"
        },
        "created_at" => "2017-01-01T00:00:00",
        "customer_hash" => "X123",
        "customer_id" => 3,
        "first_name" => "sshaw",
        "last_name" => "xxx",
        "has_uncommited_changes" => false,
        "line_items" => [
          "images" => { "small" => "http://example.com/foo.webp" },
          "subscription_id" => 9999,
          "quantity" => 10,
          "shopify_product_id" => "90",
          "shopify_variant_id" => "91",
          "sku" => "SKU111",
          "title" => "Foo Product",
          "variant_title" => "Foo Variant",
          "vendor" => "Plug Depot",
          "grams" => 453,
          "price" => 100.0,
          "properties" => [
             {
               "name" => "x",
               "value" => "y"
             }
           ]
        ],
        "note" => "noted",
        "note_attributes" => [:a => 123],
        "processed_at" => "2017-01-01T00:00:00",
        "processor_name" => "sshaw",
        "scheduled_at" => "2014-01-05T00:00:00",
        "shipments_count" => 4,
        "shopify_order_id" => "5",
        "shipping_lines" => [
          "price" => "0.00",
          "code" => "Standard Shipping",
          "title" => "Standard Shipping"
        ],
        "status" => "SUCCESS",
        "total_price" => 1.0,
        "tax_lines" => 0,
        "total_discounts" => "0.0",
        "total_line_items_price" => "12.00",
        "total_price" => "12.00",
        "total_refunds" => nil,
        "total_tax" => 0,
        "total_weight" => 4536,
        "transaction_id" => "XX_XX",
        "type" => "RECURRING",
        "updated_at" => "2017-01-03T00:00:00"
      }

      charge = described_class.new(data)
      expect(charge.to_h).to eq data
    end
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "recharge"
require "json"

require "rspec"
require "time"
require "webmock/rspec"

Recharge.api_key = "XXX"
TIME_INSTANCES = [Time.now, Date.today, DateTime.now].freeze

RSpec.configure do |c|
  c.include Module.new {
    def format_time(t)
      t.strftime("%Y-%m-%dT%H:%M:%S")
    end
  }
end

RSpec.shared_examples_for "a method that requires an id" do
  it "requires an id argument" do
    [nil, ""].each do |arg|
      expect { described_class.get(arg) }.to raise_error(ArgumentError, "id required")
    end
  end
end

RSpec::Matchers.define :define_const do |name|
  match do |klass|
    begin
      const = klass.const_get(name)
      value ? const == value : true
    rescue NameError
      false
    end
  end

  chain :set_to, :value

  failure_message do |actual|
    msg = "expected class #{actual} to define const #{expected}"
    msg << " with value '#{value}'" if value
    msg
  end
end


lib/recharge.rb
require "recharge/classes"
require "recharge/version"

module Recharge
  ENDPOINT = "api.rechargeapps.com".freeze
  PORT = 443
  TOKEN_HEADER = "X-Recharge-Access-Token".freeze
  USER_AGENT = "ReCharge API Client v#{VERSION} (Ruby v#{RUBY_VERSION})"

  Error = Class.new(StandardError)
  ConnectionError = Class.new(Error)

  #
  # Raised when a non-2XX HTTP response is returned or a response with
  # an error or warning property
  #
  class RequestError < Error
    attr_accessor :errors
    attr_accessor :status
    attr_accessor :meta

    def initialize(message, status, meta = nil, errors = nil)
      super message
      @status = status
      @meta = meta || {}
      @errors = errors || {}
    end
  end

  class << self
    attr_accessor :api_key
    # If +true+ output HTTP request/response to stderr. Can also be an +IO+ instance to output to.
    attr_accessor :debug
  end
end

ReCharge = Recharge


lib/recharge/tasks.rb
require "rake"
require "recharge"

namespace :recharge do
  namespace :hooks do
    task :_setup_recharge do
      abort "RECHARGE_API_KEY required" unless ENV["RECHARGE_API_KEY"]
      ReCharge.api_key = ENV["RECHARGE_API_KEY"]
    end

    desc "List webhooks for RECHARGE_API_KEY"
    task :list => :_setup_recharge do
      format = "%-5s %-25s %-80s\n"
      printf format, "ID", "Hook", "URL"

      Recharge::Webhook.list.each do |hook|
        printf format, hook.id, hook.topic, hook.address
      end
    end

    desc "Delete webhooks for RECHARGE_API_KEY"
    task :delete_all => :_setup_recharge do
      Recharge::Webhook.list.each { |hook| hook.class.delete(hook.id) }
    end

    desc "Delete the webhooks given by the ID(s) in ID for RECHARGE_API_KEY"
    task :delete => :_setup_recharge do
      ids = ENV["ID"].to_s.strip.split(",")
      abort "ID required" unless ids.any?

      ids.each do |id|
        puts "Deleting webhook #{id}"
        Recharge::Webhook.delete(id)
      end
    end

    desc "Create webhook HOOK to be sent to CALLBACK for RECHARGE_API_KEY"
    task :create => :_setup_recharge do
      known_hooks = %w[
        subscription/created
        subscription/updated
        subscription/activated
        subscription/cancelled
        customer/created
        customer/updated
        order/created
        charge/created
        charge/paid
      ]

      abort "CALLBACK required" unless ENV["CALLBACK"]
      abort "HOOK required" unless ENV["HOOK"]
      abort "unknown hook #{ENV["HOOK"]}" unless known_hooks.include?(ENV["HOOK"])

      puts "Creating webhook #{ENV["HOOK"]} for #{ENV["CALLBACK"]}"

      hook = Recharge::Webhook.create(
        :topic   => ENV["HOOK"],
        :address => ENV["CALLBACK"]
      )

      puts "Created hook ##{hook.id}"
    end
  end
end


lib/recharge/classes.rb
# frozen_string_literal: true

require "json"
require "class2"
require "recharge/http_request"

class2 "Recharge", JSON.parse(<<-END) do
{
  "subscription":
     {
        "id":10101,
        "address_id":178918,
        "analytics_data": {},
        "customer_id":1438,
        "cancellation_reason": "$$$$",
        "cancellation_reason_comments": "",
        "cancelled_at":"2017-02-28T20:31:29",
        "charge_interval_frequency": "3",
        "created_at":"2017-02-28T20:31:29",
        "updated_at":"2017-02-28 20:31:29",
        "next_charge_scheduled_at":"2017-04-01T00:00:00",
        "cancelled_at":null,
        "has_queued_charges": 1,
        "is_skippable": true,
        "is_swappable": true,
        "is_prepaid": false,
        "max_retries_reached": 0,
        "product_title":"Sumatra Coffee",
        "price":12,
        "quantity":1,
        "recharge_product_id": 518676,
        "status":"ACTIVE",
        "shopify_product_id":1255183683,
        "shopify_variant_id":3844924611,
        "sku_override": false,
        "sku":null,
        "order_interval_unit":"day",
        "order_interval_frequency":"30",
        "charge_interval_frequency":"30",
        "order_day_of_month":null,
        "order_day_of_week":null,
        "properties": [],
        "variant_title": "Milk - a / b"
     },
  "charge":
    {
      "address_id":178918,
      "analytics_data": {},
      "billing_address":{
         "address1":"3030 Nebraska Avenue",
         "address2":"",
         "city":"Los Angeles",
         "company":"",
         "country":"United States",
         "first_name":"Mike",
         "last_name":"Flynn",
         "phone":"",
         "province":"California",
         "zip":"90404"
      },
      "client_details": {
        "browser_ip": null,
        "user_agent": null
      },
      "created_at":"2017-03-01T19:52:11",
      "customer_hash":null,
      "customer_id":10101,
      "first_name":"Mike",
      "has_uncommited_changes": false,
      "id":1843,
      "last_name":"Flynn",
      "line_items":[
         {
            "grams":0,
            "images": {},
            "price":100.0,
            "properties":[],
            "quantity":1,
            "shopify_product_id": "1255183683",
            "shopify_variant_id":"3844924611",
            "sku":"",
            "title": "Milk  10% Off Auto renew",
            "variant_title": "a / b",
            "vendor": "Example Storeeeeeee",
            "subscription_id":14562
         }
      ],
      "note": "blah blah",
      "note_attributes": [],
      "processor_name": "foo",
      "processed_at":"2014-11-20T00:00:00",
      "scheduled_at":"2014-11-20T00:00:01",
      "shipments_count":null,
      "shipping_address":{
         "address1":"3030 Nebraska Avenue",
         "address2":"",
         "city":"Los Angeles",
         "company":"",
         "country":"United States",
         "first_name":"Mike",
         "last_name":"Flynn",
         "phone":"3103843698",
         "province":"California",
         "zip":"90404"
      },
      "shipping_lines": [],
      "shopify_order_id":"281223307",
      "status":"SUCCESS",
      "total_price":446.00,
      "updated_at":"2016-09-05T09:19:29",
      "tax_lines": 0,
      "total_discounts": "0.0",
      "total_line_items_price": "12.00",
      "total_price": "12.00",
      "total_refunds": null,
      "total_tax": 0,
      "total_weight": 4536,
      "transaction_id": "cch_1Du2QpJ2iqHvZRd18RyqoPvc",
      "type": "RECURRING"
    },
  "customer":
     {
        "id": 1438,
        "hash": "143806234a9ff87a8d9e",
        "shopify_customer_id": null,
        "email": "mike@gmail.com",
        "created_at": "2018-01-10T11:00:00",
        "updated_at": "2017-01-11T13:16:19",
        "first_name": "Mike",
        "last_name": "Flynn",
        "billing_first_name": "Mike",
        "billing_last_name": "Flynn",
        "billing_company": null,
        "billing_address1": "3030 Nebraska Avenue",
        "billing_address2": null,
        "billing_zip": "90404",
        "billing_city": "Los Angeles",
        "billing_province": "California",
        "billing_country": "United States",
        "billing_phone": "3103843698",
        "processor_type": null,
        "status": "FOO",
        "stripe_customer_token": "123123",
        "paypal_customer_token": "123123",
        "braintree_customer_token": "123123"
    },
  "order": {
     "id":7271806,
     "customer_id":10101,
     "address_id":178918,
     "charge_id":9519316,
     "transaction_id":"ch_19sdP2J2zqHvZRd1hqkeGANO",
     "shopify_order_id":"5180645510",
     "shopify_order_number":5913,
     "created_at":"2017-03-01T14:46:26",
     "updated_at":"2017-03-01T14:46:26",
     "scheduled_at":"2017-03-01T00:00:00",
     "processed_at":"2017-03-01T14:46:26",
     "status":"SUCCESS",
     "charge_status":"SUCCESS",
     "type":"CHECKOUT",
     "first_name":"Mike",
     "last_name":"Flynn",
     "email":"mike@gmail.com",
     "payment_processor":"stripe",
     "address_is_active":1,
     "is_prepaid":false,
     "line_items":[
        {
           "subscription_id":10101,
           "shopify_product_id":"1255183683",
           "shopify_variant_id":"3844924611",
           "variant_title":"Sumatra",
           "title":"Sumatra Latte",
           "quantity":1,
           "properties":[]
        }
     ],
     "total_price":18.00,
     "shipping_address":{
        "address1":"1933 Manning",
        "address2":"204",
        "city":"los angeles",
        "province":"California",
        "first_name":"mike",
        "last_name":"flynn",
        "zip":"90025",
        "company":"bootstrap",
        "phone":"3103103101",
        "country":"United States"
     },
     "billing_address":{
        "address1":"1933 Manning",
        "address2":"204",
        "city":"los angeles",
        "province":"California",
        "first_name":"mike",
        "last_name":"flynn",
        "zip":"90025",
        "company":"bootstrap",
        "phone":"3103103101",
        "country":"United States"
     }
  },
  "metafield": {
    "created_at": "2018-11-05T12:59:30",
    "description": "desc lorem ipsum",
    "id": 15,
    "key": "marjan",
    "namespace": "nmsp2c",
    "owner_id": 17868054,
    "owner_resource": "customer",
    "updated_at": "2018-11-05T15:48:42",
    "value": "5",
    "value_type": "integer"
  },
  "product": {
    "collection_id": null,
    "created_at": "2019-11-07T11:36:19",
    "discount_amount": 15.0,
    "discount_type": "percentage",
    "handle": null,
    "id": 1327844,
    "images": {},
    "product_id": 4354268856408,
    "shopify_product_id": 4354268856408,
    "subscription_defaults": {
      "charge_interval_frequency": 4,
      "cutoff_day_of_month": null,
      "cutoff_day_of_week": null,
      "expire_after_specific_number_of_charges": null,
      "modifiable_properties": [],
      "number_charges_until_expiration": null,
      "order_day_of_month": 0,
      "order_day_of_week": null,
      "order_interval_frequency": 4,
      "order_interval_frequency_options": [],
      "order_interval_unit": "month",
      "storefront_purchase_options": "subscription_only"
    },
    "title": "T-shirt",
    "updated_at": "2019-11-07T14:04:52"
  },
  "webhook": {
     "id":6,
     "address":"https://request.in/foo",
     "topic":"order/create"
  },
  "address":{
    "id":3411137,
    "address1":"1933 Manning",
    "address2":"204",
    "city":"los angeles",
    "province":"California",
    "first_name":"mike",
    "last_name":"flynn",
    "zip":"90025",
    "company":"bootstrap",
    "phone":"3103103101",
    "country":"United States"
  },
  "discount":{
    "id": 3748296,
    "code": "Discount1",
    "value": 12.5,
    "ends_at": "2019-12-15T00:00:00",
    "starts_at": "2018-05-16T00:00:00",
    "status": "enabled",
    "usage_limit": 10,
    "applies_to_id": null,
    "discount_type": "percentage",
    "applies_to": null,
    "applies_to_resource": null,
    "times_used": 0,
    "duration": "usage_limit",
    "duration_usage_limit": 10,
    "applies_to_product_type": "ALL",
    "created_at": "2018-04-25T14:32:39",
    "updated_at": "2018-05-04T13:33:53",
    "once_per_customer": false
  }
}
END
  def meta=(meta)
    @meta = meta
  end

  def meta
    @meta ||= {}
  end

  private

  def self.instance(response)
    args = response[self::SINGLE]
    args["meta"] = response["meta"]
    new(args)
  end
end

module Recharge
  module Persistable  # :nodoc:
    def save
      data = to_h
      data.delete("id")

      if id
        self.class.update(id, data)
      else
        self.id = self.class.create(data).id
      end
    end
  end

  class Address
    PATH = "/addresses".freeze
    SINGLE = "address".freeze
    COLLECTION = "addresses".freeze

    extend HTTPRequest::Get
    extend HTTPRequest::Update

    #
    # Persist the updated address
    #
    # === Errors
    #
    # Recharge::ConnectionError, Recharge::RequestError
    #
    def save
      data = to_h
      data.delete("id")
      self.class.update(id, data)
    end

    # Validate an address
    #
    # === Arguments
    #
    # [data (Hash)] Address to validate, see: https://developer.rechargepayments.com/?shell#validate-address
    #
    # === Returns
    #
    # [Hash] Validated and sometimes updated address
    #
    # === Errors
    #
    # Recharge::ConnectionError, Recharge::RequestError
    #
    # If the address is invalid a Recharge::RequestError is raised. The validation
    # errors can be retrieved via Recharge::RequestError#errors
    #
    def self.validate(data)
      POST(join("validate"), data)
    end
  end

  class Customer
    PATH = "/customers".freeze
    SINGLE = "customer".freeze
    COLLECTION = "customers".freeze

    extend HTTPRequest::Create
    extend HTTPRequest::Get
    extend HTTPRequest::Update
    extend HTTPRequest::List
    extend HTTPRequest::Count

    include Persistable

    # Retrieve all of a customer's addresses
    #
    # === Arguments
    #
    # [id (Fixnum)] Customer ID
    #
    # === Errors
    #
    # ConnectionError, RequestError
    #
    # === Returns
    #
    # [Array[Recharge::Address]] The customer's addresses
    #
    def self.addresses(id)
      id_required!(id)
      data = GET(join(id, Address::COLLECTION))
      (data[Address::COLLECTION] || []).map do |d|
        address = Address.new(d)
        address.meta = data["meta"]
        address
      end
    end

    # Create a new address
    #
    # === Arguments
    #
    # [id (Fixnum)] Customer ID
    # [address (Hash)] Address attributes, see: https://developer.rechargepayments.com/?shell#create-address
    #
    # === Returns
    #
    # [Recharge::Address] The created address
    #
    # === Errors
    #
    # Recharge::ConnectionError, Recharge::RequestError
    #
    def self.create_address(id, address)
      id_required!(id)
      data = POST(join(id, Address::COLLECTION), address)
      address = Address.new(data[Address::SINGLE])
      address.meta = data["meta"]
      address
    end
  end

  class Charge
    PATH = "/charges".freeze
    SINGLE = "charge".freeze
    COLLECTION = "charges".freeze

    extend HTTPRequest::Count
    extend HTTPRequest::Get
    extend HTTPRequest::List

    def self.list(options = nil)
      super(convert_date_params(options, :date_min, :date_max))
    end

    def self.change_next_charge_date(id, date)
      path = join(id, "change_next_charge_date")
      instance(POST(path, :next_charge_date => date_param(date)))
    end

    def self.skip(id, subscription_id)
      path = join(id, "skip")
      instance(POST(path, :subscription_id => subscription_id))
    end
  end

  class Discount
    PATH = "/discounts".freeze
    SINGLE = "discount".freeze
    COLLECTION = "discounts"

    extend HTTPRequest::Count
    extend HTTPRequest::Create
    extend HTTPRequest::Delete
    extend HTTPRequest::Get
    extend HTTPRequest::List
    extend HTTPRequest::Update

    include Persistable

    def self.count(options = nil)
      super(convert_date_params(options, :created_at_max, :created_at_min, :date_min, :date_max))
    end

    def self.list(options = nil)
      super(convert_date_params(options, :created_at, :created_at_max, :updated_at, :updated_at_max))
    end

    def delete
      self.class.delete(id)
      true
    end
  end

  class Metafield
    PATH = "/metafields"
    SINGLE = "metafield"
    COLLECTION = "metafields"

    extend HTTPRequest::Count
    extend HTTPRequest::Create
    extend HTTPRequest::Delete
    extend HTTPRequest::Get
    extend HTTPRequest::List

    include Persistable

    def self.list(owner, options = nil)
      raise ArgumentError, "owner resource required" if owner.nil? || owner.to_s.strip.empty?
      super (options||{}).merge(:owner_resource => owner)
    end

    def delete
      self.class.delete(id)
      true
    end
  end

  class Order
    PATH = "/orders".freeze
    SINGLE = "order".freeze
    COLLECTION = "orders".freeze

    extend HTTPRequest::Count
    extend HTTPRequest::Get
    extend HTTPRequest::List

    def self.count(options = nil)
      super(convert_date_params(options, :created_at_max, :created_at_min, :date_min, :date_max))
    end

    def self.change_date(id, date)
      id_required!(id)
      instance(POST(join(id, "change_date"), :shipping_date => date_param(date)))
    end

    def self.update_shopify_variant(id, old_variant_id, new_varient_id)
      id_required!(id)
      path = join(id, "update_shopify_variant", old_variant_id)
      instance(POST(path, :new_shopify_variant_id => new_varient_id))
    end
  end

  class Product
    PATH = "/products".freeze
    SINGLE = "product".freeze
    COLLECTION = "products".freeze

    extend HTTPRequest::Count
    extend HTTPRequest::Get
    extend HTTPRequest::List
  end

  class Subscription
    PATH = "/subscriptions".freeze
    SINGLE = "subscription".freeze
    COLLECTION = "subscriptions".freeze

    extend HTTPRequest::Create
    extend HTTPRequest::Get
    extend HTTPRequest::Update
    extend HTTPRequest::List

    include Persistable

    #
    # Activate a subscription
    #
    # === Arguments
    #
    # [id (Integer)] ID of subscription to cancel
    #
    # === Returns
    #
    # [Recharge::Subscription] The activated subscription
    #
    # === Errors
    #
    # Recharge::ConnectionError, Recharge::RequestError
    #
    # If the subscription was already activated a Recharge::RequestError will be raised
    #
    def self.activate(id)
      id_required!(id)
      instance(POST(join(id, "activate"), :status => "active"))
    end

    #
    # Cancel a subscription
    #
    # === Arguments
    #
    # [id (Integer)] ID of subscription to cancel
    # [reason (String)] Reason for the cancellation
    #
    # === Returns
    #
    # [Recharge::Subscription] The canceled subscription
    #
    # === Errors
    #
    # Recharge::ConnectionError, Recharge::RequestError
    #
    # If the subscription was already canceled a Recharge::RequestError will be raised
    #
    def self.cancel(id, reason)
      id_required!(id)
      instance(POST(join(id, "cancel"), :cancellation_reason => reason))
    end

    def self.set_next_charge_date(id, date)
      id_required!(id)
      instance(POST(join(id, "set_next_charge_date"), :date => date_param(date)))
    end

    def self.list(options = nil)
      #options[:status] = options[:status].upcase if options[:status]
      super(convert_date_params(options, :created_at, :created_at_max, :updated_at, :updated_at_max))
    end
  end

  class Webhook
    PATH = "/webhooks".freeze
    COLLECTION = "webhooks".freeze
    SINGLE = "webhook".freeze

    extend HTTPRequest::Create
    extend HTTPRequest::Delete
    extend HTTPRequest::Get
    extend HTTPRequest::List
    extend HTTPRequest::Update

    include Persistable

    def delete
      self.class.delete(id)
      true
    end
  end
end
